File: DynamicGraph.cpp
#include "DynamicGraph.h"

#include <cmath>
#include <utility>
#include <limits>
#include <iostream>

namespace dgraph {
    void swap(vector<Edge*>& adj, unsigned i, unsigned j) {
        if (i != j) {
            std::swap(adj[i], adj[j]);
            adj[i]->change_pos(j, i);
            adj[j]->change_pos(i, j);
        }
    }

    DynamicGraph::DynamicGraph(unsigned n) : n(n) {
        size = std::lround(std::ceil(std::log2(n)) + 1);
        for (unsigned i = 0; i < size; i++) {
            forests.emplace_back(n);
            adjLists.emplace_back();
            threshold.emplace_back();
            for (unsigned j = 0; j < n; j++){
                adjLists[i].emplace_back();
                threshold[i].push_back(0);

            }
        }
    }

    DynamicGraph::~DynamicGraph() {
        for (unsigned i = 0; i < size; i++) {
            for (unsigned j = 0; j < n; j++) {
                for (Edge* e : adjLists[i][j]) {
                    delete e;
                }
            }
        }
    }

    void DynamicGraph::add_links(Edge* e) {
        unsigned v = e->v;
        unsigned u = e->u;
        unsigned n = e->lvl;
        unsigned vs = adjLists[n][v].size();
        unsigned us = adjLists[n][u].size();
        e->subscribe(vs, us);
        adjLists[n][v].push_back(e);
        adjLists[n][u].push_back(e);
        if (e->is_tree_edge()) {
            // tree edges should be in front of adj list
            swap(adjLists[n][v], vs, threshold[n][v]++);
            swap(adjLists[n][u], us, threshold[n][u]++);
        }

    }

    EdgeToken DynamicGraph::add(unsigned v, unsigned u) {
        if (v == u) {
            return EdgeToken(nullptr);
        }
        unsigned n = size - 1;
        auto* edge = new Edge(n, v, u);
        add_links(edge);
        forests[n].increment_edges(v);
        forests[n].increment_edges(u);
        if (!is_connected(v, u)) {
            edge->add_tree_edge(forests[n].link(v, u));
        }
        return EdgeToken(edge);
    }

    void DynamicGraph::remove(EdgeToken&& edge_token) {
        Edge* link = edge_token.edge;
        if (link == nullptr) {
            return;
        }
        unsigned v = link->from();
        unsigned u = link->to();
        bool complex_deletion = link->is_tree_edge();
        unsigned level = link->level();

        if (complex_deletion) {
            for (unsigned i = 0; i <= size - level - 1; i++){
                forests[size - i - 1].cut(std::move(link->tree_edges[i]));
            }
        }

        forests[level].decrement_edges(v);
        forests[level].decrement_edges(u);

        remove_links(link);

        if (complex_deletion) {
            for (unsigned i = level; i < size; i++){
                // find new connection
                // to do that choose less component
                if(forests[i].size(v) > forests[i].size(u)){
                    std::swap(v, u);
                }
                // and iterate over good vertices until success
                // propagating all tree edges of smallest component
                Edge* replacement = nullptr;
                Iterator it = forests[i].iterator(v);
                while(it.hasNext()){
                    unsigned w = *it;
                    auto& list = adjLists[i][w];
                    for (unsigned j = 0; j < list.size(); j++) {
                        if (replacement != nullptr && i == threshold[i][w]) {
                            break;
                        }
                        Edge* e = adjLists[i][w][j];
                        unsigned up = e->opposite(w);
                        if (is_connected(up, u)) {
                            replacement = e;
                        } else {
                            downgrade(e);
                            j--;
                        }
                    }
                    ++it;
                }

                if (replacement != nullptr) {
                    // Make tree edge be in the right place in adj
                    unsigned v = replacement->v;
                    unsigned u = replacement->u;
                    swap(adjLists[i][v], replacement->first_pos, threshold[i][v]++);
                    swap(adjLists[i][u], replacement->second_pos, threshold[i][u]++);
                    for (unsigned j = size - 1; j >= i; j--){
                        replacement->add_tree_edge(forests[j].link(replacement->v, replacement->u));
                    }
                    break;
                }
            }
        }
    }

    void DynamicGraph::downgrade(Edge* e){
        unsigned v = e->from();
        unsigned w = e->to();
        remove_links(e);
        unsigned lvl = e->lvl--;
        add_links(e);
        forests[lvl].decrement_edges(w);
        forests[lvl].decrement_edges(v);
        forests[lvl - 1].increment_edges(w);
        forests[lvl - 1].increment_edges(v);
        if (e->is_tree_edge()) {
            e->add_tree_edge(forests[lvl - 1].link(v, w));
        }
    }

    bool DynamicGraph::is_connected(unsigned v, unsigned u) {
        return forests[forests.size() - 1].is_connected(v, u);
    }

    bool DynamicGraph::is_connected() {
        return forests[forests.size() - 1].is_connected();
    }

    std::string DynamicGraph::str() {
        std::string str;
        for(unsigned i = 0; i < size; i++){
            str += "level " + std::to_string(i) + ": \n";
            str += forests[i].str() + "\n";
        }
        return str;
    }

    unsigned DynamicGraph::degree(unsigned v) {
        unsigned sum = 0;
        for (unsigned i = 0; i < size; i++) {
            sum += forests[i].degree(v);
        }
        return sum;
    }

    void DynamicGraph::remove_links(Edge* e) {
        unsigned v = e->v;
        unsigned u = e->u;
        unsigned lvl = e->lvl;
        auto& vs = adjLists[lvl][v];
        auto& us = adjLists[lvl][u];
        if (e->is_tree_edge()) {
            swap(vs, e->first_pos, threshold[lvl][v] - 1);
            swap(us, e->second_pos, threshold[lvl][u] - 1);
        }
        swap(vs, e->first_pos, vs.size() - 1);
        swap(us, e->second_pos, us.size() - 1);
        vs.pop_back();
        us.pop_back();
    }

    Edge::Edge(unsigned lvl, unsigned v, unsigned u) : lvl(lvl), v(v), u(u) {}

    void Edge::subscribe(unsigned first, unsigned second) {
        first_pos = first;
        second_pos = second;
    }

    unsigned Edge::level() {
        return lvl;
    }

    unsigned Edge::from() {
        return v;
    }

    unsigned Edge::to() {
        return u;
    }

    void Edge::add_tree_edge(TreeEdge&& edge) {
        tree_edges.push_back(std::move(edge));
    }

    bool Edge::is_tree_edge() {
        return !tree_edges.empty();
    }

    void Edge::change_pos(unsigned vertex, unsigned pos) {
        if (u == vertex) {
            first_pos = pos;
        } else {
            second_pos = pos;
        }
    }

    unsigned Edge::opposite(unsigned w) {
        return v = w ? u : v;
    }

    EdgeToken::EdgeToken(Edge* edge) :edge(edge){}

    EdgeToken::EdgeToken(EdgeToken&& e) noexcept :edge(e.edge){
        e.edge = nullptr;
    }

    EdgeToken& EdgeToken::operator=(EdgeToken&& other) noexcept {
        edge = other.edge;
        other.edge = nullptr;
        return *this;
    }

    EdgeToken::EdgeToken() :edge(nullptr){}
}

File: java/src/main/java/ctlab/dgraph/DynamicGraph.java
package ctlab.dgraph;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * This class implements an algorithm for fully dynamic connectivity. It allows to add edges, remove edges and
 * check connectivity between arbitrary vertices as well as check that overall graph is connected. The only limitation
 * is that number of vertices must be fixed at the moment of object construction.
 *
 * @author Alexander Loboda
 */
public class DynamicGraph {
    private int n;
    private int size;
    private List<EulerTourForest> forests;
    private List<List<LinkedList>> adjLists;

    /**
     * Constructs a new dynamic graph with n vertices. Vertices are numbered from 0 to n - 1.
     * @param n Number of vertices.
     */
    public DynamicGraph(int n) {
        if (n <= 0) {
            throw new IllegalArgumentException();
        }
        this.n = n;
        size = (int)Math.round(Math.ceil(Math.log(n) / Math.log(2)) + 1);
        forests = new ArrayList<>();
        adjLists = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            forests.add(new EulerTourForest(n));
            adjLists.add(new ArrayList<>());
            for (int j = 0; j < n; j++) {
                adjLists.get(i).add(new LinkedList());
            }
        }
    }

    /**
     * Returns true if a path connecting vertices v and u exists.
     * @param v a vertex
     * @param u a vertex
     * @return boolean
     */
    public boolean isConnected(int v, int u) {
        return forests.get(forests.size() - 1).isConnected(v, u);
    }

    /**
     * Returns true if all n vertices are connected.
     * @return boolean
     */
    public boolean isConnected() {
        return forests.get(forests.size() - 1).isConnected();
    }

    /**
     * Adds an edge between vertices v and u to the graph.
     * @param v a vertex
     * @param u a vertex
     * @return a token that could be used to remove the edge from the graph
     */
    public EdgeToken add(int v, int u) {
        if (v == u) {
            return new EdgeToken(null);
        }
        if (v >= this.n || u >= this.n || v < 0 || u < 0) {
            throw new IllegalArgumentException("No such vertex in the graph");
        }
        int n = size - 1;
        Edge edge = new Edge(n, v, u);
        if (!isConnected(v, u)) {
            edge.addTreeEdge(forests.get(n).link(v, u));
        }
        forests.get(n).incrementEdges(v);
        forests.get(n).incrementEdges(u);
        edge.subscribe(adjLists.get(n).get(v).add(u, edge),
                       adjLists.get(n).get(u).add(v, edge));
        return new EdgeToken(edge);
    }

    /**
     * Removes previously added edge from the graph. The operation requires a token of corresponding addition operation.
     * Applying this method twice on the same token has no effect.
     * @param token a token
     */
    public void remove(EdgeToken token) {
        Edge link = token.edge;
        token.edge = null;
        if (link == null) {
            return;
        }

        int v = link.from();
        int u = link.to();
        boolean complex_deletion = link.isTreeEdge();
        int level = link.level();

        if (complex_deletion) {
            for (int i = 0; i <= size - level - 1; i++){
                forests.get(size - i - 1).cut(link.treeEdges.get(i));
            }
        }

        forests.get(level).decrementEdges(v);
        forests.get(level).decrementEdges(u);

        link.removeLinks();

        if (complex_deletion) {
            for (int i = level; i < size; i++){
                // find new connection
                // to do that choose the lesser component
                if(forests.get(i).size(v) > forests.get(i).size(u)){
                    int t = v;
                    v = u;
                    u = t;
                }
                // and iterate over good vertices until success
                // propagating all tree edges of smallest component
                Edge replacement = null;
                Iterator<Integer> it = forests.get(i).iterator(v);
                while(it.hasNext()){
                    Integer w = it.next();
                    Iterator<LinkedList> lit = adjLists.get(i).get(w).iterator();
                    while(lit.hasNext()){
                        LinkedList l = lit.next();
                        Edge e = l.edge;
                        int up = l.vertex();
                        if (e.isTreeEdge()) {
                            downgrade(e);
                        } else {
                            if (replacement != null) {
                                break;
                            }
                            if (isConnected(up, u)) {
                                replacement = e;
                            } else {
                                downgrade(e);
                            }
                        }
                    }
                }

                if (replacement != null) {
                    for (int j = size - 1; j >= i; j--){
                        replacement.addTreeEdge(forests.get(j).link(replacement.v, replacement.u));
                    }
                    break;
                }
            }
        }
    }

    void downgrade(Edge e){
        int v = e.from();
        int w = e.to();
        int lvl = e.lvl--;
        e.removeLinks();
        e.subscribe(adjLists.get(lvl - 1).get(w).add(v, e), adjLists.get(lvl - 1).get(v).add(w, e));
        forests.get(lvl).decrementEdges(w);
        forests.get(lvl).decrementEdges(v);
        forests.get(lvl - 1).incrementEdges(w);
        forests.get(lvl - 1).incrementEdges(v);
        if (e.isTreeEdge()) {
            e.addTreeEdge(forests.get(lvl - 1).link(v, w));
        }
    }

    public static class EdgeToken {
        private Edge edge;

        EdgeToken(Edge edge) {
            this.edge = edge;
        }
    }

    static class Edge {
        private int lvl;
        private int v;
        private int u;
        private LinkedList firstLink;
        private LinkedList secondLink;
        private List<EulerTourForest.TreeEdge> treeEdges;

        Edge(int lvl, int v, int u) {
            this.lvl = lvl;
            this.v = v;
            this.u = u;
            treeEdges = new ArrayList<>();
        }

        void subscribe(LinkedList first, LinkedList second) {
            firstLink = first;
            secondLink = second;
        }

        int level() {
            return lvl;
        }

        void removeLinks() {
            firstLink.remove();
            secondLink.remove();
            firstLink = null;
            secondLink = null;
        }

        int from() {
            return v;
        }

        int to() {
            return u;
        }

        void addTreeEdge(EulerTourForest.TreeEdge edge) {
            treeEdges.add(edge);
        }

        boolean isTreeEdge() {
            return !treeEdges.isEmpty();
        }
    }

    static class LinkedList {
        private Edge edge;
        private int u;
        private LinkedList next;
        private LinkedList prev;

        LinkedList(int u, Edge edge, LinkedList prev, LinkedList next) {
            this.u = u;
            this.edge = edge;
            this.next = next;
            this.prev = prev;
        }

        LinkedList add(int v, Edge edge) {
            LinkedList newList = new LinkedList(v, edge, prev, this);
            prev.next = newList;
            prev = newList;
            return newList;
        }

        void remove() {
            next.prev = prev;
            prev.next = next;
        }

        LinkedList() {
            next = this;
            prev = this;
            u = Integer.MAX_VALUE;
        }

        Iterator<LinkedList> iterator() {
            return new ListIterator(next);
        }

        int vertex() {
            return u;
        }

        Edge e() {
            return edge;
        }
    }

    static class ListIterator implements Iterator<LinkedList> {
        private LinkedList list;

        ListIterator(LinkedList list) {
            this.list = list;
        }

        @Override
        public boolean hasNext() {
            return list.edge != null;
        }

        @Override
        public LinkedList next() {
            LinkedList ret = list;
            list = list.next;
            return ret;
        }
    }
}
